python/hsfs/transformation_function.py

```diff
@@ -0,0 +1,262 @@
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+
+import humps
+import json
+import inspect
+import ast
+
+from hsfs import util, engine
+from hsfs.core import transformation_function_engine
+
+
+class TransformationFunction:
+    UDF = "@udf"
+    PANDAS_UDF = "@pandas_udf"
+    PYSPARK_TYPE_IMPORT = (
+        "from pyspark.sql.functions import udf\nfrom pyspark.sql.types import *"
+    )
+
+    def __init__(
+        self,
+        featurestore_id,
+        transformation_fn=None,
+        version=None,
+        name=None,
+        source_code_content=None,
+        output_type=None,
+        id=None,
+        type=None,
+        items=None,
+        count=None,
+        href=None,
+    ):
+        self._id = id
+        self._featurestore_id = featurestore_id
+        self._version = version
+        self._name = name
+        self._transformation_fn = transformation_fn
+        self._source_code_content = source_code_content
+
+        self._transformation_function_engine = (
+            transformation_function_engine.TransformationFunctionEngine(
+                self._featurestore_id
+            )
+        )
+
+        # set up depending on user initialized or coming from backend response
+        if self._transformation_fn is not None:
+            # type -> user init coming from user
+            self._transformer_code = None
+            self._extract_source_code()
+            self._output_type = self._transformation_function_engine.infer_spark_type(
+                output_type
+            )
+        else:
+            # load original source code
+            self._output_type = output_type
+            self._load_source_code(json.loads(self._source_code_content))
+
+    def save(self):
+        """Persist transformation function in backend."""
+        self._transformation_function_engine.save(self)
+
+    def delete(self):
+        """Delete transformation function from backend."""
+        self._transformation_function_engine.delete(self)
+
+    def _extract_source_code(self):
+        if not callable(self._transformation_fn):
+            raise ValueError("transformer must be callable")
+
+        self._name = self._transformation_fn.__name__
+
+        transformer_code = inspect.getsource(self._transformation_fn)
+
+        module_imports = self._get_module_imports(
+            self._get_module_path(self._transformation_fn.__module__)
+        )
+
+        self._transformer_code = "\n".join(module_imports) + "\n" + transformer_code
+
+        # initialise source code dict
+        # add all imports from module
+        # add original source code that will be used during offline transformations
+        self._source_code_content = json.dumps(
+            {
+                "module_imports": "\n".join(module_imports),
+                "transformer_code": transformer_code,
+            }
+        )
+
+    @staticmethod
```

Comment:
You don't need the `@staticmethod` annotation when you call methods with `self.`

Reply:
Its fine for me but in that case style check shows warning